drivers/audio/wm8904.c

```diff
@@ -0,0 +1,728 @@
+		wm8904_set_master_clock(dev, &cfg->dai_cfg, cfg->mclk_freq);
+	} else {
+		/* BCLK/LRCLK default direction input */
+		wm8904_update_reg(dev, WM8904_REG_AUDIO_IF_1, 1U << 6U, 0U);
+		wm8904_update_reg(dev, WM8904_REG_AUDIO_IF_3, (uint16_t)(1UL << 11U), 0U);
+	}
+
+	switch (cfg->dai_route) {
+	case AUDIO_ROUTE_PLAYBACK:
+		wm8904_configure_output(dev);
+		break;
+
+	case AUDIO_ROUTE_CAPTURE:
+		wm8904_configure_input(dev);
+		break;
+
+	case AUDIO_ROUTE_PLAYBACK_CAPTURE:
+		wm8904_configure_output(dev);
+		wm8904_configure_input(dev);
+		break;
+
+	default:
+		break;
+	}
+
+	return 0;
+}
+
+static void wm8904_start_output(const struct device *dev)
+{
+}
+
+static void wm8904_stop_output(const struct device *dev)
+{
+}
+
+static int wm8904_set_property(const struct device *dev, audio_property_t property,
+			       audio_channel_t channel, audio_property_value_t val)
+{
+	switch (property) {
+	case AUDIO_PROPERTY_OUTPUT_VOLUME:
+		return wm8904_out_volume_config(dev, channel, val.vol);
+
+	case AUDIO_PROPERTY_OUTPUT_MUTE:
+		return wm8904_out_mute_config(dev, channel, val.mute);
+
+	case AUDIO_PROPERTY_INPUT_VOLUME:
+		return wm8904_in_volume_config(dev, channel, val.vol);
+
+	case AUDIO_PROPERTY_INPUT_MUTE:
+		return wm8904_in_mute_config(dev, channel, val.mute);
+	}
+
+	return -EINVAL;
+}
+
+static int wm8904_apply_properties(const struct device *dev)
+{
+	/**
+	 * Set VU = 1 for all output channels, VU takes effect for the whole
+	 * channel pair.
+	 */
+	wm8904_update_reg(
+		dev,
+		WM8904_REG_ANALOG_OUT1_LEFT,
+		WM8904_REGVAL_OUT_VOL(0, 1, 0, 0),
+		WM8904_REGMASK_OUT_MUTE
+	);
+	wm8904_update_reg(dev,
+		WM8904_REG_ANALOG_OUT2_LEFT,
+		WM8904_REGVAL_OUT_VOL(0, 1, 0, 0),
+		WM8904_REGMASK_OUT_MUTE
+	);
+
+	return 0;
+}
+
+static void wm8904_write_reg(const struct device *dev, uint8_t reg, uint16_t val)
+{
+	const struct wm8904_driver_config *const dev_cfg = DEV_CFG(dev);
+	uint8_t data[3];
+	int ret;
+
+	/* data is reversed */
+	data[0] = reg;
+	data[1] = (val >> 8) & 0xff;
+	data[2] = val & 0xff;
+
+	ret = i2c_write(dev_cfg->i2c.bus, data, 3, dev_cfg->i2c.addr);
+
+	if (ret != 0) {
+		LOG_ERR("i2c write to codec error %d", ret);
+	}
+
+	LOG_DBG("REG:%02u VAL:%#02x", reg, val);
+}
+
+static void wm8904_read_reg(const struct device *dev, uint8_t reg, uint16_t *val)
+{
+	const struct wm8904_driver_config *const dev_cfg = DEV_CFG(dev);
+	uint16_t value;
+	int ret;
+
+	ret = i2c_write_read(dev_cfg->i2c.bus, dev_cfg->i2c.addr, &reg, sizeof(reg), &value,
+			     sizeof(value));
+	if (ret == 0) {
+		*val = (value >> 8) & 0xff;
+		*val += ((value & 0xff) << 8);
+		/* update cache*/
+		LOG_DBG("REG:%02u VAL:%#02x", reg, *val);
+	}
+}
+
+static void wm8904_update_reg(const struct device *dev, uint8_t reg, uint16_t mask, uint16_t val)
+{
+	uint16_t reg_val = 0;
+	uint16_t new_value = 0;
+
+	if (reg == 0x19) {
+		LOG_INF("try write mask %#x val %#x", mask, val);
```

Comment:
These `LOG_INF()` calls in this function will probably generate a lot of logging messages. Wouldn't `LOG_DBG()` be a better choice?